cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java

```diff
@@ -200,4 +244,104 @@ public void getAppDetailForVersion(HttpRequest request, HttpResponder responder,
         : applicationLifecycleService.getAppDetail(appId);
     responder.sendJson(HttpResponseStatus.OK, GSON.toJson(appDetail));
   }
+
+  @PUT
+  @Path("/apps/{app-id}")
+  @AuditPolicy(AuditDetail.REQUEST_BODY)
+  public BodyConsumer create(HttpRequest request, HttpResponder responder,
+      @PathParam("namespace-id") final String namespaceId,
+      @PathParam("app-id") final String appId,
+      @QueryParam("skipMakingLatest") final boolean skipMakingLatest)
+      throws BadRequestException, NamespaceNotFoundException, AccessException {
+    String versionId = ApplicationId.DEFAULT_VERSION;
+    if (Feature.LIFECYCLE_MANAGEMENT_EDIT.isEnabled(featureFlagsProvider)) {
+      versionId = RunIds.generate().getId();
+    }
+    ApplicationId applicationId = validateApplicationVersionId(validateNamespace(namespaceId), appId, versionId);
+
+    try {
+      return deployAppFromArtifact(applicationId, skipMakingLatest);
+    } catch (Exception ex) {
+      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR,
+          "Deploy failed: " + ex.getMessage());
+      return null;
+    }
+  }
+
+  private BodyConsumer deployAppFromArtifact(final ApplicationId appId, final boolean skipMakingLatest)
```

Comment:
I had considered to put it in the abstract class. But there was a minor difference between the internal and public implementations (in case of access validations). Will try and refactor it.

Reply:
Just adding a TODO with a [JIRA](https://cdap.atlassian.net/browse/CDAP-20847) task for now. Will take it up in the AppLifecycleService cleanup PR.

Reply:
The question was never answered, is this duplicate code or is it different some how? If it is different, please highlight what is changed and why. If it's the same, we should just move it now and not have a todo.

Reply:
moved to the base class and updated usage.